File: workstation/engine/src/main/java/io/xj/engine/fabricator/MemeIsometry.java
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.
package io.xj.engine.fabricator;

import io.xj.hub.entity.EntityException;
import io.xj.hub.entity.EntityUtils;
import io.xj.hub.meme.Isometry;
import io.xj.hub.meme.MemeStack;
import io.xj.hub.meme.MemeTaxonomy;
import io.xj.hub.util.StringUtils;

import java.util.Collection;
import java.util.List;
import java.util.Objects;

/**
 Determine the isometry between a source and target group of Memes
 */
public class MemeIsometry extends Isometry {
  static final String KEY_NAME = "name";
  final MemeStack stack;

  /**
   Construct a meme isometry from source memes

   @param sourceMemes from which to construct isometry
   */
  MemeIsometry(MemeTaxonomy taxonomy, Collection<String> sourceMemes) {
    for (String meme : sourceMemes) add(StringUtils.toMeme(meme));
    stack = MemeStack.from(taxonomy, getSources());
  }

  /**
   Instantiate a new MemeIsometry of a group of source Memes

   @param taxonomy    context within which to measure isometry
   @param sourceMemes to compare of
   @return MemeIsometry ready for comparison to target Memes
   */
  public static MemeIsometry of(MemeTaxonomy taxonomy, Collection<String> sourceMemes) {
    return new MemeIsometry(taxonomy, sourceMemes);
  }

  /**
   Instantiate a new MemeIsometry representing having no memes

   @return an empty MemeIsometry
   */
  public static MemeIsometry none() {
    return new MemeIsometry(MemeTaxonomy.empty(), List.of());
  }

  /**
   Score a CSV list of memes based on isometry to source memes

   @param targets comma-separated values to score against source meme names
   @return score is between 0 (no matches) and the number of matching memes
   */
  public int score(Collection<String> targets) {
    if (!isAllowed(targets)) return 0;
    return targets.stream()
      .map(StringUtils::toMeme)
      .flatMap(target -> sources.stream().map(source ->
        Objects.equals(source, target) ? 1 : 0))
      .reduce(0, Integer::sum);
  }

  /**
   Add a meme for isometry comparison
   */
  public <R> void add(R meme) {
    try {
      EntityUtils.get(meme, KEY_NAME)
        .ifPresent(name -> add(StringUtils.toMeme(String.valueOf(name))));
    } catch (EntityException ignored) {
    }
  }

  public boolean isAllowed(Collection<String> memes) {
    return stack.isAllowed(memes);
  }
}

File: engine/src/meme/MemeIsometry.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <sstream>
#include <unordered_set>

#include "xjmusic/meme/MemeIsometry.h"

using namespace XJ;

MemeIsometry::MemeIsometry(MemeTaxonomy taxonomy, const std::set<std::string> &sourceMemes)
    : stack(std::move(taxonomy), sourceMemes) {
  for (const std::string &meme: sourceMemes) add(StringUtils::toMeme(meme));
}

MemeIsometry MemeIsometry::of(MemeTaxonomy taxonomy, const std::set<std::string> &sourceMemes) {
  return MemeIsometry(std::move(taxonomy), sourceMemes);
}

MemeIsometry MemeIsometry::of(const std::set<std::string> &sourceMemes) {
  return MemeIsometry(MemeTaxonomy::empty(), sourceMemes);
}

MemeIsometry MemeIsometry::none() {
  return MemeIsometry(MemeTaxonomy(), std::set<std::string>());
}

int MemeIsometry::score(const std::set<std::string> &targets) const {
  if (!isAllowed(targets)) return 0;

  int sum = 0;
  for (const auto &target: targets) {
    std::string targetMeme = StringUtils::toMeme(target);
    for (const auto &source: sources) {
      if (source == targetMeme) {
        sum += 1;
      }
    }
  }

  return sum;
}

void MemeIsometry::add(const std::string &meme) {
  sources.insert(StringUtils::toMeme(meme));
}

void MemeIsometry::add(const ProgramMeme &meme) {
  add(meme.name);
}

void MemeIsometry::add(const ProgramSequenceBindingMeme &meme) {
  add(meme.name);
}

void MemeIsometry::add(const InstrumentMeme &meme) {
  add(meme.name);
}

void MemeIsometry::add(const SegmentMeme &meme) {
  add(meme.name);
}

bool MemeIsometry::isAllowed(const std::set<std::string> &memes) const {
  return stack.isAllowed(memes);
}

std::set<std::string> MemeIsometry::getSources() {
  return sources;
}

std::string MemeIsometry::getConstellation() const {
  return MemeConstellation::fromNames(sources);
}


const std::string MemeIsometry::KEY_NAME = "name";

